gcp-autocomplete-lib.js

```diff
@@ -0,0 +1,107 @@
+const GCCompute = require("@google-cloud/compute");
+const { ProjectsClient } = require("@google-cloud/resource-manager");
+const _ = require("lodash");
+
+const {
+  RESOURCE_OPERATIONS,
+} = require("./gcp-lib");
+const parsers = require("./parsers");
+const helpers = require("./helpers");
+
+function createListItemsFunction(clientClass, fields) {
+  return async function listGcpItems(query, pluginSettings, pluginParams) {
+    const { result } = await getListResults(fields, pluginSettings, pluginParams, clientClass);
+    return result;
+  };
+}
+
+async function listGcpProjects(query, pluginSettings, pluginParams) {
+  const { result } = await getProjects(["projectId", "displayName"], pluginSettings, pluginParams);
+  return result;
+}
+
+async function listGcpRegions(query, pluginSettings, pluginParams) {
+  const { result } = await getListResults(["name"], pluginSettings, pluginParams, GCCompute.RegionsClient);
+  return result;
+}
+
+async function listGcpZones(query, pluginSettings, pluginParams) {
+  const { result, params } = await getListResults(["name"], pluginSettings, pluginParams, GCCompute.ZonesClient);
+
+  const region = parsers.autocomplete(params.region);
+  result.filter((zone) => !region || zone.name.includes(region));
+
+  return result;
+}
+
+async function getProjects(fields, pluginSettings, pluginParams) {
+  const parseFunc = helpers.getParseFromParam(...fields);
+  const settings = helpers.mapAutoParams(pluginSettings);
+  const params = helpers.mapAutoParams(pluginParams);
+
+  let result = await searchProjects(params, settings);
+  result = [...result.map(parseFunc)];
+  return { params, result };
```

Comment:
Why return params if they're not used anywhere?

Reply:
Follow up: I think I know what you meant to do here – pass the parsed parameters to the caller for eventual further usage. I don't like this flow though. `getProjects` should give you projects only, parsed parameters as a part of returned value does not make sense in this context. Maybe consider parsing the parameters outside the `getProjects` (and other) functions and handle the parsing at the callers' level?

Reply:
Redesigned